routing/endpointregistry.go

```diff
@@ -83,57 +81,35 @@ func NewEndpointRegistry(o RegistryOptions) *EndpointRegistry {
 	}
 
 	return &EndpointRegistry{
-		data:            map[string]*entry{},
-		lastSeen:        map[string]time.Time{},
+		data:            sync.Map{},
 		lastSeenTimeout: o.LastSeenTimeout,
 		now:             time.Now,
 	}
 }
 
 func (r *EndpointRegistry) GetMetrics(key string) Metrics {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-
-	e := r.getOrInitEntryLocked(key)
+	e, _ := r.data.LoadOrStore(key, &entry{})
 	copy := &entry{}
-	*copy = *e
+	*copy = *e.(*entry)
 	return copy
 }
 
 func (r *EndpointRegistry) SetDetectedTime(key string, detected time.Time) {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-
-	e := r.getOrInitEntryLocked(key)
-	e.detected = detected
+	e, _ := r.data.LoadOrStore(key, &entry{})
+	e.(*entry).detected = detected
```

Comment:
It is not safe to read and write this field concurrently. I think we use it only in post-processor but nothing tells client that it should not be used elsewhere.

Maybe we can make it non-exported or introduce a constructor like

 ```go
 func (r *EndpointRegistry) GetOrCreateMetrics(key string) (m Metrics, created bool) 
 ```

Reply:
> Maybe we can make it non-exported

I think I can make it non-exported (after changing some test code a little bit), it's not a problem.

> or introduce a constructor like

I don't understand how creating additional return value in getter can help us avoid get+set of the same timestamp races.

Reply:
The idea was that GetOrCreateMetrics would set detected time and created would hint if metrics were created or reused similar to LoadOrStore

Reply:
BTW don't we need to allow SetDetectedTime for the endpointCreated filter?

Reply:
> BTW don't we need to allow SetDetectedTime for the endpointCreated filter?

https://opensource.zalando.com/skipper/reference/filters/#endpointcreated
> This filter is typically automatically appended

AFAIR endpointCreated is needed to store the information about endpoint detection timestamp and needed only for us as Skipper developers, not for Skipper users. As endpointregistry becomes available, I don't see any point for us as Skipper developers to have this filter, because endpointregistry stores the information itself.
Should we still give users an options to use `endpointCreated`?

Reply:
I decided to keep `setDetectedTime` non-exported, it's easy to make export it if required.